src/Analyzer/Passes/DuplicateOrderByPass.cpp

```diff
@@ -0,0 +1,115 @@
+{
+public:
+    bool is_stateful = false;
+
+    bool needChildVisit(VisitQueryTreeNodeType &, VisitQueryTreeNodeType &) const { return !is_stateful; }
+
+    void visitImpl(VisitQueryTreeNodeType & node)
+    {
+        if (!node)
+            return;
+
+        if (node->getNodeType() != QueryTreeNodeType::FUNCTION)
+            return;
+
+        const auto * function_node = node->as<FunctionNode>();
+
+        auto aggregate_function_properties = AggregateFunctionFactory::instance().tryGetProperties(function_node->getFunctionName());
+        if (aggregate_function_properties && aggregate_function_properties->is_order_dependent)
+        {
+            is_stateful = true;
+            return;
+        }
+
+        if (function_node->isOrdinaryFunction() && function_node->getFunction()->isStateful())
+        {
+            is_stateful = true;
+            return;
+        }
+    }
+};
+
+class DeduplicateOrderByVisitor : public InDepthQueryTreeVisitor<DeduplicateOrderByVisitor>
+{
+    bool try_drop_order_by_in_subquery = false;
+    bool an_upper_query_break_order = false;
+
+    static void tryEliminateOrderBy(QueryNode * query_node)
+    {
+        if (!query_node->hasOrderBy())
+            return;
+
+        /// If we have limits then the ORDER BY is non-removable
+        if (query_node->hasLimitBy() || query_node->hasLimitByLimit() || query_node->hasLimitByOffset())
+            return;
+
+        /// If ORDER BY contains filling (in addition to sorting) it is non-removable.
+        auto & order_by_nodes = query_node->getOrderBy().getNodes();
+        for (const auto & sort_node : order_by_nodes)
+        {
+            const auto & sort_node_typed = sort_node->as<const SortNode &>();
+            if (sort_node_typed.withFill())
+                return;
+        }
+
+        query_node->getOrderByNode()->getChildren().clear();
+    }
+
+public:
+    void visitImpl(VisitQueryTreeNodeType & node)
+    {
+        auto * query_node = node->as<QueryNode>();
+        if (!query_node)
+            return;
+
+        /// drop ORDER BY in current subquery if necessary
+        if (query_node->isSubquery() && an_upper_query_break_order)
+            tryEliminateOrderBy(query_node);
+
+        /// check if we can eliminate ORDER BY in subquery
+        const auto & projection_nodes = query_node->getProjection().getNodes();
+        for (const auto & elem : projection_nodes)
+        {
+            FunctionStatefulVisitor function_visitor;
+            function_visitor.visit(elem);
+            if (function_visitor.is_stateful)
+            {
+                try_drop_order_by_in_subquery = false;
+                return;
+            }
+        }
+
+        if (an_upper_query_break_order)
+            try_drop_order_by_in_subquery = true;
+
+        if (!an_upper_query_break_order)
+            an_upper_query_break_order = query_node->hasOrderBy() || query_node->hasGroupBy();
```

Comment:
I didn't get what is incorrect here. The example, you've provided, has no case to apply the optimization, and the implementation doesn't change the corresponding query tree.

Reply:
Yes, my example is not the case when this optimization applies. I wanted to show you that the value of `an_upper_query_break_order` is propagated not only into the subtree but also into all the other nodes.

When your visit `t1` you will set `an_upper_query_break_order = true` and if `t2` contains `ORDER BY` this optimization can be applied even if an outer query doesn't have `ORDER BY` or `GROUP BY`.

This example is not very good because we don't need an order for joins but the purpose was to illustrate the issue.